### PR TITLE
C#: Add generated higher order models for .NET8 Runtime.

### PR DESCRIPTION
**Pull request by michaelnebel**
I spot checked some of the added LINQ models. They look good. Some models related to *key* *comparing* is missing, but I don't think this is due to the higher order model generation (I think this is related to flow for dictionaries).

DCA looks good, but unfortunately, we don't find any new results.

### CODE REVIEW

**Review by hvitved — APPROVED** [review]
Very nice! The few summaries that I checked look great. Will be interesting to see what QA says.

File: csharp/ql/lib/ext/generated/System.Collections.model.yml
Comment (tamasvajk) [inline csharp/ql/lib/ext/generated/System.Collections.model.yml:53]: Do these summaries make sense? I think generally `Equals` would not propagate taint from arg to qualifier. Similarly I don't think `GetHashCode` should propagate flow either. There could be implementations that cache the arguments, and therefore taint is propagated to the qualifier, but that behavior doesn't match the general purpose of these methods.

File: csharp/ql/lib/ext/generated/System.Collections.model.yml
Comment (tamasvajk) [inline csharp/ql/lib/ext/generated/System.Collections.model.yml:45]: The same comment applies as with `System.Collections.IEqualityComparer`

File: csharp/ql/lib/ext/generated/Microsoft.DotNet.PlatformAbstractions.model.yml
Comment (tamasvajk) [inline csharp/ql/lib/ext/generated/Microsoft.DotNet.PlatformAbstractions.model.yml:7]: I'd expect an `Argument[0]` to `Argument[this]` summary here.

**Comment by michaelnebel** [comment]
> Very nice! The few summaries that I checked look great. Will be interesting to see what QA says.

Do you think it is worth running a QA experiment before merging this branch?

File: csharp/ql/lib/ext/generated/System.Reflection.Emit.model.yml
Comment (tamasvajk) [inline csharp/ql/lib/ext/generated/System.Reflection.Emit.model.yml:175]: Are the removals in this file expected?

**Comment by hvitved** [comment]
> Do you think it is worth running a QA experiment before merging this branch?

I think it's fine to wait until the next release QA run.

**Review by tamasvajk — COMMENTED** [review]
Spot checked a couple of models. I highlighted some that look odd.

File: csharp/ql/lib/ext/generated/System.Collections.model.yml
Comment (michaelnebel) [reply csharp/ql/lib/ext/generated/System.Collections.model.yml:53]: No, these are not very good summaries. The provenance of the models are `df-generated`, so these models come from the "old" heuristic for detecting models.
In this case, exactly as you guess, because there exists an implementation, where the heuristic derives that flow propgates to `this` (as there is a field containing a `Func` which is applied to the arguments) and then the model is lifted.
One could argue yet another improvement we could consider is *only* to lift models in case the models are *functional* - but I think that is out of scope for this PR.

File: csharp/ql/lib/ext/generated/System.Collections.model.yml
Comment (michaelnebel) [reply csharp/ql/lib/ext/generated/System.Collections.model.yml:45]: I haven't checked this concretely, but with very high probability it is the same issue as mentioned above.

File: csharp/ql/lib/ext/generated/Microsoft.DotNet.PlatformAbstractions.model.yml
Comment (michaelnebel) [reply csharp/ql/lib/ext/generated/Microsoft.DotNet.PlatformAbstractions.model.yml:7]: The model generator considers simple types as sanitisers. The backing field containing the hash code is a `long`, so we shouldn't expect the `Argument[0] -> Argument[this]` model.
However, the reason we are getting the `Argument[0] -> Argument[1]` (bad) model is for the same reason that we get the `IEqualityComparer<TValue>.Compare` models (there exists an implementation that we can dispatch to, which is considered to manipulate the state of `this` for the `IEqualityComparer`).

File: csharp/ql/lib/ext/generated/System.Reflection.Emit.model.yml
Comment (michaelnebel) [reply csharp/ql/lib/ext/generated/System.Reflection.Emit.model.yml:175]: Yes, these removals are of *neutrals* and we don't generate a neutral since we are generating a model for a base implementation of this method
```
 - ["System.Reflection", "MethodBase", True, "Invoke", "(System.Object,System.Reflection.BindingFlags,System.Reflection.Binder,System.Object[],System.Globalization.CultureInfo)", "", "Argument[0]", "Argument[this]", "taint", "df-generated"]
```

**Comment by michaelnebel** [comment]
@tamasvajk : Thank you for the review - I have tried to answer the questions the best I can. Thank you for pinpointing some good examples for further inspection. I don't think there are any of these cases that requires that we make an immediate improvement before releasing the models in their current state.

**Review by tamasvajk — APPROVED** [review]

**Pull request merged**
